### ISSUE TITLE
Release strategy / breaking changes

### ISSUE THREAD

**Issue by skolima** [System-IO-Abstractions/System.IO.Abstractions#338]
Current release strategy is somewhat difficult to follow for downstream projects. I know it's been discussed https://github.com/System-IO-Abstractions/System.IO.Abstractions/issues/295 but even finding any release notes/changelog is currently not very straightforward https://github.com/System-IO-Abstractions/System.IO.Abstractions/issues/308 .

I would like to make another point, though: System.IO.Abstractions is both a) strong-signed b) including minor and patch version number in the signature c) consisting of two separate packages (effectively - as there's no point in using just the main assembly without having tests, in most cases). This makes for quite a cumbersome exercise, trying to keep the dependency up to date, as effectively every S.IO.A. release is a breaking changes release - and having to maintain 2 separate packages in sync and updating the app.config / web.config files across all projects. And if one happens to publish nuget packages depending on System.IO.Abstractions, this whole dance gets repeated when consuming those. So...

I don't have a solution for all those pain-points I'm experiencing. One approach might be what https://github.com/JamesNK/Newtonsoft.Json is using, that is only incrementing the version in signature on major/breaking releases (it always ends with .0 and only the informational version is maintained per-build). This avoids one of the 3 problems (`.config` files), but Newtonsoft.Json doesn't ship as two separate packages.  Slowing the release cadence might be another solution, with commits to master only triggering a preview build to be published (which I understand is not what S.IO.A team would like :-) ). My personal preference would be to have an unsigned version available (judging by how much people asked for a signed version to be back when this mechanism was broken suggests this would have to be a second, separate version published).

Thanks for keeping this very useful project alive! 👍

**Comment by jpreese** [comment]
As mentioned in #295 I had a lot of the same feelings skolima. It does seem like an agressive approach to releasing. Coming from a work environment where we have many .NET applications, all with their own NuGet references, a release forces us to update the app and redeploy them all. This approach does keep it hard to stay up to date.

Though as @tathamoddie mentioned, we're in kind of a unique space where we just mimic an already public facing API, there *shouldnt* be much in terms of breaking changes. So I agree SemVer probably doesn't make a ton of sense and the currently build approach we have is OK. Eventually we're going to get to a point where no active development can occur. We'll have implemented everything that we want to in System.IO and the bugs will be gone. Once we match System.IO to the level that we want to, there's nothing left to implement.

The only fault comes back to what you mentioned. 

Do we slow our feature releases and release more than one feature per build?
- Features/bug fixex will be sitting idle waiting for other development to conclude.
- Less package versions for consumers to manage.

Do we keep it the same?
- Features/bug fixes are available as soon as they are ready
- More package versions for consumers to manage.

Honestly, we decided to stick with the latter, purely because the concern had yet to be raised by the community. But here we are, with an issue about too many packages.

We're definitely open to changing our release strategy if it becomes too much of a problem.

**Comment by ericnewton76** [comment]
@skolima you are right, and I'm actually seeing a lot more packages utilizing the 1.0.0.0 version in AssemblyVersion, which .Net fusion (the assembly binder) uses to decide to load types, versus AssemblyFileVersion and AssemblyInformationalVersion which easily helps identify, for humans, what version of the DLL it is.

And to reinforce @jpreese assertions, yes, the wrapper for System.IO definitely needs to slow down on releases, ie, the actual abstraction library, S.IO.A, should slow to a crawl.  The S.IO.Mocks theoretically can be devdependencies, but most Nugets dont have that concept so until the Nuget team gets their act straight, and start releasing Nuget properly to downlevel VS, (at least 2015, lots of enterprise customers still on that), then there's (an as-usual with Nuget) version complication with S.IO.A and Mocks.

For my musings, IMO, the Mocks library is really becoming a virtual file system, having all the bugs with none of the benefits...   The Abstraction tests should only test the actual abstractions library, and since its abstract, what is there to test?   I'm thinking the Mocks library should be separated, and allowed to grow organically, releasing when prudent. 

This keeps S.IO.A releases to a bugfix that somehow got introduced in the abstraction.

**Comment by fgreinacher** [comment]
I really love the idea of versioning the TestingHelpers separately from the Abstractions. Most issues we have (and work we do) only affects TestingHelpers. I think we we don’t need a more complicated versioning/strategy if TestingHelpers would be a package with a separate version. The only challenge would be to split stuff properly repo-/CI-wise but IMO that’d be definitely worth it!

**Comment by ericnewton76** [comment]
Splitting is easy, it'll be getting the dependency properly setup on the core S-IO-A.dll nuget package.  

We could then mark a line in the sand, where S-IO-A itself stops incrementing and stays pretty static.  

Plan would be, assuming current version is 2.1.0: AssemblyVersion gets set to `[assembly: AssemblyVersion("2.0.0.0")]` and (if not present, add) `[assembly: AssemblyInformationalVersion("2.1.1.0")`.  

Of note, Net40 added EnumerateFiles and such, so a change like that in the future warrants a version bump.

The System.IO.Mocks library then depends on S-IO-A v2.0.0.0 and the versioning issues start to melt away.  You can leave S-IO.A at version 2 and have S-IO-Mocks depend on latest nuget 2.1.2

Plan:
- [ ] Split System.IO.Mocks to separate repo
- [ ] Set System.IO.Mocks version to 2.1.2
- [ ] S-IO-Abstractions AssemblyVersion gets set to `[assembly: AssemblyVersion("2.0.0.0")]` 
- [ ] S-IO-Abstractions AssemblyInformationVersion gets set `[assembly: AssemblyInformationalVersion("2.1.2.0")`.  
- [ ] System.IO.Mocks gets dependency set to System.IO.Abstractions^2.0.0 and will end up referencing System.IO.Abstractions.dll AssemblyVersion 2.0.0.0.
- [ ] Viola

**Comment by fgreinacher** [comment]
We now changed a few things:
- the assembly version is now more stable (MAJOR.0.0.0) to reduce downstream pain
- we create a tag/release on every commit on master to help seeing differences between releases
- we still deploy every commit on master to NuGet.org, but usually those will be prerelease packages.
- the current release process is described at https://github.com/System-IO-Abstractions/System.IO.Abstractions/blob/master/CONTRIBUTING.md

The separation of TestingHelper is tracked via #381.

@skolima I'd suggest to close this issue and track other possible improvements separately. What do you think?

**Comment by skolima** [comment]
It's much easier now to consume the library, great work!
